core/src/main/java/org/apache/calcite/sql/SqlJsonTypeNameSpec.java

```diff
@@ -0,0 +1,117 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.util.Litmus;
+
+/**
+ * A sql type name specification of the JSON type.
+ */
+public class SqlJsonTypeNameSpec extends SqlTypeNameSpec {
+
+  private final Integer maxLength;
+  private final Integer inlineLength;
+  private final CharacterSet characterSet;
+  private final StorageFormat storageFormat;
+
+  public SqlJsonTypeNameSpec(SqlParserPos pos,
+      Integer maxLength,
+      Integer inlineLength,
+      CharacterSet characterSet,
+      StorageFormat storageFormat) {
+    super(new SqlIdentifier("JSON", pos), pos);
+    this.maxLength = maxLength;
+    this.inlineLength = inlineLength;
+    this.characterSet = characterSet;
+    this.storageFormat = storageFormat;
+
+    // These two fields are mutally exclusive.
+    assert characterSet == null || storageFormat == null;
+
+    if (characterSet != null) {
+      assert characterSet == CharacterSet.LATIN
+        || characterSet == CharacterSet.UNICODE;
+    }
+  }
+
+  @Override public RelDataType deriveType(SqlValidator validator) {
+    return validator.getValidatedNodeType(getTypeName());
+  }
+
+  @Override public boolean equalsDeep(SqlTypeNameSpec spec, Litmus litmus) {
+    if (!(spec instanceof SqlJsonTypeNameSpec)) {
+      return litmus.fail("{} != {}", this, spec);
+    }
+    SqlJsonTypeNameSpec that = (SqlJsonTypeNameSpec) spec;
+    if (!this.maxLength.equals(that.getMaxLength())
+        || !this.inlineLength.equals(that.getInlineLength())
+        || this.characterSet != that.getCharacterSet()
+        || this.storageFormat != that.getStorageFormat()) {
+      return litmus.fail("{} != {}", this, spec);
+    }
+    return litmus.succeed();
+  }
+
+  @Override public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
+    writer.keyword("JSON");
+    if (maxLength != null) {
+      final SqlWriter.Frame frame =
+          writer.startList(SqlWriter.FrameTypeEnum.FUN_CALL, "(", ")");
+      writer.print(maxLength);
+      writer.endList(frame);
+    }
+    if (inlineLength != null) {
+      writer.keyword("INLINE LENGTH " + inlineLength);
+    }
+    if (characterSet != null) {
+      writer.keyword("CHARACTER SET " + characterSet.toString());
+    } else if (storageFormat != null) {
+      writer.keyword("STORAGE FORMAT " + storageFormat.toString());
+    }
+  }
+
+  public Integer getMaxLength() {
```

Comment:
Per feedback I've received from Julian on the preferred convention in Calcite, let's convert this (and all other getters below) into just public final fields.